python/lsst/cp/pipe/ptc.py

```diff
@@ -369,147 +362,218 @@ def runDataRef(self, dataRef, visitPairs):
             dataRef.dataId['expId'] = v1
-            exp1 = self.isr.runDataRef(dataRef).exposure
+            exp1 = dataRef.get("postISRCCD", immediate=True)
             dataRef.dataId['expId'] = v2
-            exp2 = self.isr.runDataRef(dataRef).exposure
+            exp2 = dataRef.get("postISRCCD", immediate=True)
             del dataRef.dataId['expId']
 
             checkExpLengthEqual(exp1, exp2, v1, v2, raiseWithMessage=True)
             expTime = exp1.getInfo().getVisitInfo().getExposureTime()
 
-            for amp in detector:
-                mu, varDiff = self.measureMeanVarPair(exp1, exp2, region=amp.getBBox())
+            tupleRows = []
+            for ampNumber, amp in enumerate(detector):
                 ampName = amp.getName()
+                # covAstier: (i, j, var (cov[0,0]), cov, npix)
+                doRealSpace = self.config.covAstierRealSpace
+                muDiff, varDiff, covAstier = self.measureMeanVarCov(exp1, exp2, region=amp.getBBox(),
+                                                                    covAstierRealSpace=doRealSpace)
+
+                datasetPtc.rawExpTimes[ampName].append(expTime)
+                datasetPtc.rawMeans[ampName].append(muDiff)
+                datasetPtc.rawVars[ampName].append(varDiff)
+                datasetPtc.inputVisitPairs[ampName].append((v1, v2))
+
+                tupleRows += [(muDiff, ) + covRow + (ampNumber, expTime, ampName) for covRow in covAstier]
+                tags = ['mu', 'i', 'j', 'var', 'cov', 'npix', 'ext', 'expTime', 'ampName']
+            allTags += tags
+            tupleRecords += tupleRows
+        covariancesWithTags = np.core.records.fromrecords(tupleRecords, names=allTags)
+
+        if self.config.ptcFitType in ["ASTIERFULL", ]:
+            # Calculate covariances and fit them, including the PTC, to Astier+19 full model (Eq. 20)
+            datasetPtc = self.fitCovariancesAstier(datasetPtc, covariancesWithTags)
+        elif self.config.ptcFitType in ["ASTIERAPPROXIMATION", "POLYNOMIAL"]:
+            # Fit the PTC to a polynomial or to Astier+19 approximation (Eq. 16)
+            # Fill up PhotonTransferCurveDataset object.
+            datasetPtc = self.fitPtc(datasetPtc, self.config.ptcFitType)
+
+        # Fit a poynomial to calculate non-linearity and persist linearizer.
+        if self.config.doCreateLinearizer:
+            numberAmps = len(amps)
+            numberAduValues = self.config.maxAduForLookupTableLinearizer
+            lookupTableArray = np.zeros((numberAmps, numberAduValues), dtype=np.float32)
+
+            # Fit (non)linearity of signal vs time curve.
+            # Fill up PhotonTransferCurveDataset object.
+            # Fill up array for LUT linearizer (tableArray).
+            # Produce coefficients for Polynomial ans Squared linearizers.
+            # Build linearizer objects.
+            datasetPtc, linsArray = self.fitNonLinearityAndBuildLinearizers(datasetPtc, detector,
```

Comment:
This shouldn't modify datasetPtc.  All of the linearity information should be contained in the linearity object.  If something needs to be added to the linearity class definition to support that, we should add that.  Saving that in the PTC is not obvious to me.